### DISCUSSION TITLE
Help migrating from JWT 3.3.3 to 5.2

### DISCUSSION THREAD

**Discussion by Duque2011** [lcobucci/jwt#1050] [Q&A]
Good afternoon devs.

I'm an almost junior dev (hahaha), and I've been in the field for about a year. To make sure I don't miss any work, I'm accepting all challenges.

The latest task that came my way is to update an entire API for a system. The big issue is that it was using JWT 3.3.3, and I'm having trouble making it work on 5.2 because a lot has changed.

Can anyone help if I share a screenshot of how it's currently configured?

**Comment by SvenRtbg** [comment]
I would suggest you follow the upgrade path: Version 3.4 should be helpful (it was when I worked on a similar task some years ago), and then go to 4.x and 5.x. This may sound like a lot of useless steps, but the library tries to send helpful deprecation and change comments (turn on the error level in PHP to E_ALL for them to surface in your tests).

Also read the docs. Especially the one valid for these earlier versions. And follow the upgrading steps: https://lcobucci-jwt.readthedocs.io/en/4.4.x/upgrading/

**Reply by Duque2011** [reply]
I've been following this step-by-step guide for a few days now. However, what I tried was to upgrade directly to 5.2. I'll try following this advice and go first to version 3.4, then 4.0, and finally 5.2.

**Reply by Duque2011** [reply]
In the version 3.4, I managed to generate the token; however, I had to remove this line:

->expiresAt($time + 3600)

The original project has a time of 30 minutes. Can I add something similar?

**Reply by SvenRtbg** [reply]
I feel like the parameter you pass to the function should be a DateTimeInterface - the migration guide suggest using `->expiresAt($now->modify('+1 hour'))` with `$now = new \DateTimeImmutable`.

**Reply by Duque2011** [reply]
if($user){
                
                if($request->P_SYSTEM_ID == 1){
                    $signer = new Sha256();
                    $time = new DateTimeImmutable();
                    $token = (new Builder())
                        ->withClaim('uid', $user->ID)
                        ->issuedAt($time)
                        //->expiresAt($time)
                        ->getToken($signer, new Key(env('JWT_SECRET')));
                }else{...

**Reply by SvenRtbg** [reply]
->expiresAt($time->modify('+30 minutes'))
    
Or the token would expire immediately

**Reply by Duque2011** [reply]
Success

**Reply by SvenRtbg** [reply]
Well... it automatically starts working once you do it right, doesn't it. :)

**Reply by Duque2011** [reply]
Now I'm moving on to the next step, migrating from version 3.4 to 4.0.

**Comment by Duque2011** [comment]
<img width="780" alt="Captura de Tela 2024-02-23 às 18 40 36" src="https://github.com/lcobucci/jwt/assets/67491037/f2e4a98f-1b45-47f6-91bc-4c3110f0cbfe">


In which file should I make this change?

**Reply by Duque2011** [reply]
<img width="796" alt="Captura de Tela 2024-02-23 às 18 44 04" src="https://github.com/lcobucci/jwt/assets/67491037/15e37eb1-b108-455e-9b07-865175921224">


I'm in doubt about which of these two files I should do this.

**Reply by SvenRtbg** [reply]
It depends. The diffed file is a hypothetical example of your own code that tries to issue a token, and illustrates the necessary changes if you want to switch from providing all single parts towards configuring this config object somewhere and here using it. Note that this config object isn't magically created, and using it is definitely optional. I can't really remember if I used it - my use case was only consuming tokens created somewhere else, so I had no use to create tokens, and the config object seemed to provide no obvious benefit, instead it would ask for things like the private encryption key that I could not provide.

You have to assess what your use case is - and even if you use the library both ways, it might be your existing code is unwilling to be pushed into utilizing this config object, because you might not have central dependency injection, or whatever other reason you may encounter.

I believe migrating towards this config object is optional, if you are willing to do some steps manually. And it could be done later, once you reached 5.x.

**Reply by Duque2011** [reply]
<img width="322" alt="Captura de Tela 2024-02-23 às 19 24 29" src="https://github.com/lcobucci/jwt/assets/67491037/d90456c5-97fb-427d-bcd3-5123c4beb003">

See what I'm receiving. I think there's still something wrong.

**Reply by lcobucci** [reply]
The usage of the configuration object is indeed optional. It is useful when you have multiple places issuing and parsing tokens.

If that's not the case for you, then skip it.
